### PR TITLE
Creates generic device type testing framework

### PR DESCRIPTION
**Pull request by mruberry**
This PR addresses #24851 by...

1. lets device types easily register themselves for testing
2. lets tests be written to run on multiple devices and with multiple dtypes
3. provides a mechanism to instantiate those tests so they are discoverable and filterable by unittest and pytest

It refactors three tests from test_torch.py to demonstrate how to use it. 

`test_diagonal` is the simplest example. Most tests just need to be modified to accept 'device' as an argument. The framework will then instantiate `test_diagonal_cpu` and `test_diagonal_cuda` (when CUDA is available) which call `test_diagonal` with the appropriate 'device' argument.

`test_neg` also has dtype variants. It accepts both 'device' and 'dtype' as arguments, and the dtypes it runs with are specified with the 'dtypes' decorator. Dtypes can be specified for all device types and particular device types. The framework instantiates tests like `test_neg_cpu_torch.float`.

`test_inverse` has device-specific dependencies. These dependencies are expressed with the sugary 'skipCUDAIfNoMagma' and 'skipCPUIfNoLapack' decorators. These decorators are device-specific so CPU testing is not skipped if Magma is not installed, and there conditions may be checked after or before the test case has been initialized. This means that skipCUDAIfNoMagma does not initialize CUDA. In fact, CUDA is only initialized if a CUDA test is run.

These instantiated tests may be run as usual and with pytest filtering it's easy to run one test on all device types, run all the tests for a particular device type, or run a device type and dtype combination. 

See the note "Generic Device-Type Testing" for more detail.

### CODE REVIEW

**Comment by ezyang** [comment]
I probably won't get around to reviewing this today; hopefully tomorrow.

**Comment by mruberry** [comment]
I am preparing a fix for the Python 2 issues. There is actually a type check in Python 2 where the self parameter to a method must be the expected class. This check was removed in Python 3 since it's unPythonic.

File: test/common_device_type.py [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+# To write a test that runs on a variety of devices do the following:
+#
+#   (1) Create a class, "class <name>(<base>)"
+#   (2) Write tests as usual, except they must take a 'device' argument
+#       and may take a 'dtype' argument. If the test takes the 'dtype'
+#       argument then it must specify a set of dtypes using the @dtypes
+#       decorator. Additional decorators can specialize the set of dtypes
+#       for CPU or CUDA device types.
+#
+#       The device will be a string, like 'cpu' or 'cuda,' and dtype will be a
```
[the remark this answers is not in the document]
Comment (mruberry) [reply test/common_device_type.py]: Unfortunately no. There's not even a generic way to query for all available devices, which would have been a preferable discovery mechanism.

**Review by ezyang — CHANGES_REQUESTED** [review]
Major correctness problems:
* Use of `fn.__name__` to index into `_generic_test_dtypes`. This entire cluster of code seems a bit fishy; more on design notes below.
* You MUST NOT use dots in the generated test method names, this breaks the test runner interface

Design-wise:
* The cluster of code involving decorators is very strange and hard to understand, because (1) they are not actually wrapping and (2) they're not written in idiomatic Python decorator style. There are a number of places where decorators cause side effects, storing values into global maps. This lead to the `fn.__name__` bug, and in general I think it would be better if per-function state is stored locally, on the functions themselves, rather than in a big glob of global state. E.g., tests should know their opt in. With a decorator, the way I usually expect this sort of thing to happen is that the state is stored in the closure of the function that wraps the original function.
* dtype decorator should support exclusions, since I suspect the common path will be you only run test on one dtype (default), or you run test on all dtypes, or you run test on all dtypes except the few that don't work

**Review by gchanan — CHANGES_REQUESTED** [review]

File: test/common_device_type.py [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+        d[self.device_type] = self.args
+        setattr(fn, 'dtypes', d)
+        return fn
+
+# Sugar that requests a variant of the test for each dtype.
+def allDtypes(fn):
+    return dtypes(*all_dtypes)(fn)
+
+# Sugar that requests a variant of the test for each dtype EXCEPT those given.
+class allDtypesExcept(dtypes):
```
[the remark this answers is not in the document]
Comment (ezyang) [reply test/common_device_type.py]: @gchanan In an earlier version of this patch, there was a big hard coded list of dtypes at https://github.com/pytorch/pytorch/blob/c5b6525ffb8ee7263d5aa3826eebf4a04a166336/test/test_torch.py#L13342

I don't want there to be N copies of this long list of dtypes everywhere in the test suite. We can quibble about how exactly they should be abstracted, but there should be some abstraction.

In any case, this is mooted as long as we don't actually add the dtype decorators.

File: test/test_torch.py [hunk 3]
```diff
@@ -13451,8 +13326,126 @@ def add_neg_dim_tests():
...
+        self.assertEqual(result, expected)
+
+        x = torch.randn((100, 100), device=device)
+        result = torch.diagonal(x, 17)
+        expected = torch.diag(x, 17)
+        self.assertEqual(result, expected)
+
+    def test_neg(self, device):
+        int_types = [torch.int, torch.short, torch.int8, torch.uint8]
+        float_types = [torch.float, torch.double, torch.long]
```
Comment (ezyang) [inline test/test_torch.py]: @gchanan There's still an explicit list here!

**Review by ezyang — APPROVED** [review]
okey dokey

File: test/common_device_type.py [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+
+# Note: Generic Device-Type Testing
+#
+# [WRITING TESTS]
+#
+# Write your test class as usual except:
+#   (1) Only define test methods in the test class itself. Helper methods
+#       and non-methods must be inherited. This limitation is for Python2
+#       compatibility.
+#   (2) Each test method should have have the signature
```
Comment (gchanan) [inline test/common_device_type.py]: nit: one have

File: test/common_device_type.py [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+# In short, if you write a test signature like
+#   def textX(self, device)
+# You are effectively writing
+#   def testX_cpu(self, device='cpu')
+#   def textX_cuda(self, device='cuda')
+#   def testX_xla(self, device='xla')
+#   ...
+#
+# These tests can be run directly like normal tests:
+# "python test_torch.py TestTorchDeviceTypeCPU.test_diagonal_cpu"
```
Comment (gchanan) [inline test/common_device_type.py]: are we blocked on just calling these "TestTorchCPU" and similar?

File: test/common_device_type.py [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+if torch.cuda.is_available():
+    device_type_test_bases.append(CUDATestBase)
+
+
+# Adds 'instantiated' device-specific test cases to the given scope.
+# The tests in these test cases are derived from the generic tests in
+# generic_test_class.
+# See note "Generic Device Type Testing."
+def instantiate_device_type_tests(generic_test_class, scope):
+    # Removes the generic test class from its enclosing scope so it's tests
```
Comment (gchanan) [inline test/common_device_type.py]: nit: its

File: test/test_torch.py [hunk 7]
```diff
@@ -13451,8 +13326,126 @@ def add_neg_dim_tests():
         assert not hasattr(_TestTorchMixin, test_name), "Duplicated test name: " + test_name
         setattr(_TestTorchMixin, test_name, make_neg_dim_test(name, tensor_arg, arg_constr, types, extra_dim))
 
-add_neg_dim_tests()
+# Device-agnostic tests. Instantiated below and not run directly.
```
Comment (gchanan) [inline test/test_torch.py]: err, agnostic isn't the word I would use for these tests.  Device-generic?  Generic-Device-aware?

File: test/test_torch.py (hunk 3 above)
Comment (gchanan) [reply test/test_torch.py]: also interesting that `float_types` has `long` in the list?

Comment (gchanan) [reply test/test_torch.py]: which runs randn?  how can this possibly work?

File: test/test_torch.py [hunk 8]
```diff
@@ -13451,8 +13326,126 @@ def add_neg_dim_tests():
...
+
+            # test out of place as well
+            res_neg_out_place = a.clone().neg()
+            self.assertEqual(res_neg_out_place, res_add)
+
+            # test via __neg__ operator
+            res_neg_op = -a.clone()
+            self.assertEqual(res_neg_op, res_add)
+
+    @skipCUDAIfNoMagma
```
Comment (gchanan) [inline test/test_torch.py]: can I parameterize these based on device?

**Review by gchanan — APPROVED** [review]

File: test/common_device_type.py (hunk 5 above)
Comment (mruberry) [reply test/common_device_type.py]: We can call them whatever you like but there already is a distinct TestTorch class. Maybe there won't be after we refactor?

File: test/test_torch.py (hunk 7 above)
Comment (mruberry) [reply test/test_torch.py]: Device-generic sounds good.

File: test/test_torch.py (hunk 8 above)
Comment (mruberry) [reply test/test_torch.py]: Sorry, what do you mean?

File: test/common_device_type.py (hunk 5 above)
Comment (gchanan) [reply test/common_device_type.py]: good question, but there isn't a TestTorchCPU class in any case :).

File: test/test_torch.py (hunk 8 above)
Comment (gchanan) [reply test/test_torch.py]: I mean I'd rather write something like:
`@skipIf(device == 'CUDA' and not TEST_MAGMA)`
than
`@skipCUDAIfNotMagma`

I'm just concerned we are going to have a lot of these decorators when they are all easily expressed with code snippets (assuming we can parameterize based on device).

Comment (mruberry) [reply test/test_torch.py]: Oh I see. Yes we could support syntax like that. You'd have to put the reason in, too:

@skipIf(device == 'cuda' and not TEST_MAGMA, "no MAGMA library detected")

In practice I don't think there are many device-specific decorators, though.

**Pull request closed without merging**